exec/java-exec/src/main/java/org/apache/drill/exec/physical/rowSet/impl/ResultVectorCache.java

```diff
@@ -0,0 +1,181 @@
+ * <li>If the set of vectors changes in any way (add a vector, remove a
+ * vector, change the type of a vector), then the <tt>next()</tt> call
+ * <b>must</b> return <tt>OK_NEW_SCHEMA</tt>.</ul>
+ * </ul>
+ * These rules create interesting constraints for the scan operator.
+ * Conceptually, each batch is distinct. But, it must share vectors. The
+ * {@link ResultSetLoader} class handles this by managing the set of vectors
+ * used by a single reader.
+ * <p>
+ * Readers are independent: each may read a distinct schema (as in JSON.)
+ * Yet, the Drill protocol requires minimizing spurious <tt>OK_NEW_SCHEMA</tt>
+ * events. As a result, two readers run by the same scan operator must
+ * share the same set of vectors, despite the fact that they may have
+ * different schemas and thus different <tt>ResultSetLoader</tt>s.
+ * <p>
+ * The purpose of this inventory is to persist vectors across readers, even
+ * when, say, reader B does not use a vector that reader A created.
+ * <p>
+ * The semantics supported by this class include:
+ * <ul>
+ * <li>Ability to "pre-declare" columns based on columns that appear in
+ * an explicit select list. This ensures that the columns are known (but
+ * not their types).</li>
+ * <li>Ability to reuse a vector across readers if the column retains the same
+ * name and type (minor type and mode.)</li>
+ * <li>Ability to flush unused vectors for readers with changing schemas
+ * if a schema change occurs.</li>
+ * <li>Support schema "hysteresis"; that is, the a "sticky" schema that
+ * minimizes spurious changes. Once a vector is declared, it can be included
+ * in all subsequent batches (provided the column is nullable or an array.)</li>
+ * </ul>
+ */
+public class ResultVectorCache {
+
+  /**
+   * State of a projected vector. At first all we have is a name.
+   * Later, we'll discover the type.
+   */
+
+  private static class VectorState {
+    protected final String name;
+    protected ValueVector vector;
+    protected boolean touched;
+
+    public VectorState(String name) {
+      this.name = name;
+    }
+
+    public boolean satisfies(MaterializedField colSchema) {
+      if (vector == null) {
+        return false;
+      }
+      MaterializedField vectorSchema = vector.getField();
+      return vectorSchema.getType().equals(colSchema.getType());
+    }
+  }
+
+  private final BufferAllocator allocator;
+  private final Map<String, VectorState> vectors = new HashMap<>();
+
+  public ResultVectorCache(BufferAllocator allocator) {
+    this.allocator = allocator;
+  }
+
+  public void predefine(List<String> selected) {
+    for (String colName : selected) {
+      addVector(colName);
+    }
+  }
+
+  private VectorState addVector(String colName) {
+    VectorState vs = new VectorState(colName);
+    vectors.put(vs.name, vs);
+    return vs;
+  }
+
+  public void newBatch() {
+    for (VectorState vs : vectors.values()) {
+      vs.touched = false;
+    }
+  }
+
+  public void trimUnused() {
```

Comment:
Was one attempt to handle changing schemas by removing vectors that where created in one batch, but then not used in a later batch. The original design of some of Drill's code enforced the idea that required vectors are required. So, if a required vector is added, but not data written to it, what to do? The thought here was to remove the vector.

Later, it became clear that the only workable solution is to support "fill-empties" even for required vectors: we must fill unused vectors with zeros. It then becomes up to the client to determine if this makes sense; the underlying framework no longer has an opinion.